src/types.ts

```diff
@@ -1,8 +1,79 @@
 import * as mongoose from 'mongoose';
 
-import type { Base } from './defaultClasses';
+import type { Base, TimeStamps } from './defaultClasses';
 import type { Severity, WhatIsIt } from './internal/constants';
 
+// tslint:disable:ban-types
+
+type IfEquals<X, Y, A, B> =
+  (<T>() => T extends X ? 1 : 2) extends
+  (<T>() => T extends Y ? 1 : 2) ? A : B;
+
+type ReadonlyKeysOf<T> = {
+  [P in keyof T]: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, never, P>
+}[keyof T];
+
+type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? K : never }[keyof T];
+
+export interface TypegooseModel<
+  T extends mongoose.Document,
+  QueryHelpers = {},
+  V = Omit<T,
+    ReadonlyKeysOf<T>
+    | FunctionPropertyNames<T>
+    | keyof mongoose.Document
+    | keyof IObjectWithTypegooseFunction
+    | (T extends TimeStamps ? keyof TimeStamps : never)> & Partial<Pick<mongoose.Document, '_id' | '__v'>>,
+  D = { [K in keyof V]: V[K] extends Map<infer X, infer Y> ?
+    X extends string | number | symbol ? Record<X, Y> | Map<X, Y> : Map<X, Y> : V[K] }>
+  extends Omit<Pick<mongoose.Model<T, QueryHelpers>, keyof mongoose.Model<T, QueryHelpers>>, 'create'> {
+  /**
+   * Model constructor
+   * Provides the interface to MongoDB collections as well as creates document instances.
+   * @param doc values with which to create the document
+   * @event error If listening to this event, it is emitted when a document
+   *   was saved without passing a callback and an error occurred. If not
+   *   listening, the event bubbles to the connection used to create this Model.
+   * @event index Emitted after Model#ensureIndexes completes. If an error
+   *   occurred it is passed with the event.
+   * @event index-single-start Emitted when an individual index starts within
+   *   Model#ensureIndexes. The fields and options being used to build the index
+   *   are also passed with the event.
+   * @event index-single-done Emitted when an individual index finishes within
+   *   Model#ensureIndexes. If an error occurred it is passed with the event.
+   *   The fields, options, and index name are also passed.
+   */
+  new <ExtraOmittedKeys extends keyof D = never, P extends Omit<D, ExtraOmittedKeys> = Omit<D, ExtraOmittedKeys>>(doc?: P): T;
+  /**
+   * Shortcut for saving one or more documents to the database. MyModel.create(docs)
+   * does new MyModel(doc).save() for every doc in docs.
+   * Triggers the save() hook.
+   */
+  create<ExtraOmittedKeys extends keyof D = never, P extends Omit<D, ExtraOmittedKeys> = Omit<D, ExtraOmittedKeys>>(docs: P[],
+    callback?: (err: any, res: T[]) => void): Promise<T[]>;
```

Comment:
should we remove the callback structure from the typegoose-mongoose types? (typegoose is more or less centered around using Promises)

Reply:
I agree, since this is a breaking change anyway, we might as well push best practices.

Reply:
completely removed since 2d920dea63953c2ed8b63de6cea7c754d8bdda63